### PR TITLE
single, global xrefs dict instead of many lists

### PR DESCRIPTION
**Pull request by radare**
This patch was originally written by @rlaemmert 

r2r PR: https://github.com/radare/radare2-regressions/pull/1278

### CODE REVIEW

**Comment by radare** [comment]
let's wait for travis, and ill try to find some time to finish the patch.

**Comment by radare** [comment]
rebased and removed debug messages

**Comment by XVilka** [comment]
@radare shall be this merged?

**Comment by radare** [comment]
This branch cannot be rebased due to conflicts

**Comment by radare** [comment]
ive fixed the git mess, let's wait for travis again

**Comment by radare** [comment]
25 XX, lets move for 2.6, but pretty close

**Comment by ret2libc** [comment]
Also, I know it isn't strictly part of this PR, but it's the first time listxrefs function is used (since before USE_DICT was 0), and I don't like at all how the `dict` data structure is iterated. It requires knowledge of how it is implemented internally and I have not/shouldn't have any idea how it works internally.

**Comment by ret2libc** [comment]
I think I found the problem of the above issue:
```
diff --git a/libr/anal/xrefs.c b/libr/anal/xrefs.c
index 20478c5ae..8e6089002 100644
--- a/libr/anal/xrefs.c
+++ b/libr/anal/xrefs.c
@@ -80,9 +80,8 @@ static void listrefs(dict *m, ut64 addr, RList *list) {
                for (i = 0; i < m->size; i++) {
                        dictkv *kv = m->table[i];
                        if (kv) {
-                               dict *ht = kv->u;
                                while (kv->k != MHTNO) {
-                                       mylistrefs (ht, UT64_MAX, list);
+                                       mylistrefs (kv->u, UT64_MAX, list);
                                        kv++;
                                }
                        }
@@ -114,9 +113,8 @@ static void listxrefs(dict *m, ut64 addr, RList *list) {
                for (i = 0; i < m->size; i++) {
                        dictkv *kv = m->table[i];
                        if (kv) {
-                               dict *ht = kv->u;
                                while (kv->k != MHTNO) {
-                                       mylistrefs (ht, UT64_MAX, list);
+                                       mylistrefs (kv->u, UT64_MAX, list);
                                        kv++;
                                }
                        }

```

**Comment by ret2libc** [comment]
Please push fixed tests to the refs-dict branch in r2r (https://github.com/radare/radare2-regressions/pull/1278)

**Comment by ret2libc** [comment]
As things are implemented now, this is wrong. The dictionary of XREFS/REFS has just an address as a key, so if you have multiple `call`s to the same function, just one will be shown because the other one will be overwritten in the dictionary.

I think the solution is to have `(from, to)` as key of the XREFS/REFS dictionaries.

**Comment by radare** [comment]
Nope. As i said in the irc

**Comment by ret2libc** [comment]
There were bugs in the implementation. Right now, xrefs is a dict of dict, but in both entries the key is `from`, whether instead `from` should be the key only of the top-level dict, while `to` should be the key of the second dictionary (this actually is equivalent to a `(from,to`) key of a single dictionary).

**Comment by ret2libc** [comment]
This PR is introducing *a lot* of overhead when doing analysis (which requires xrefs). Dict is not a real hash table because it doesn't grow and it needs to be eliminated IMO.

**Comment by ret2libc** [comment]
The other source of problems is that when doing analysis, we often need the Xrefs/refs of a function and that needs to be dynamically reconstructed each time.

**Comment by XVilka** [comment]
@ret2libc do you have the numbers? to compare times/loops before/after?

**Comment by ret2libc** [comment]
@XVilka I can get the numbers. I'll provide them as soon as i'll look again at the performances. I'm trying to reach green tests before that.

@radare One issue here is that since we are using an hashmap, we don't have ordered xrefs... the easiest solution may be to just sort the xrefs when printing. It shouldn't introduce too much overhead and it's just when printing.

**Comment by ret2libc** [comment]
@XVilka about numbers:
master:
```
r2 ./bins/anal/keygen.exe
Warning: off > bin->size
Warning: off > bin->size
 -- I nodejs so hard my exams. What a nodejs!
[0x004c6c7c]> ?t aa
[x] Analyze all flags starting with sym. and entry0 (aa)
0.002024
[0x004c6c7c]> ?t aac
2.749659
[0x004c6c7c]> ?t aar
2.373480
[0x004c6c7c]> ?t aan
0.017311
```
refs-dict:
```
$ r2 ./bins/anal/keygen.exe
Warning: off > bin->size
Warning: off > bin->size
 -- WASTED
[0x004c6c7c]> ?t aa
[x] Analyze all flags starting with sym. and entry0 (aa)
0.002270
[0x004c6c7c]> ?t aac
7.324354
[0x004c6c7c]> ?t aar
1.321755
[0x004c6c7c]> ?t aan
72.293789
```

I already have some optimizations in mind for the `aan` command and also for the others. I will implement them shortly.

**Comment by radare** [comment]
Sgtm

**Comment by ret2libc** [comment]
New results after `core/canal.c: avoid iterating fcn xrefs list` commit.

```
r2 ./radare2-regressions/bins/anal/keygen.exe
Warning: off > bin->size
Warning: off > bin->size
 -- r2 is meant to be read by machines.
[0x004c6c7c]> ?t aa
[x] Analyze all flags starting with sym. and entry0 (aa)
0.002011
[0x004c6c7c]> ?t aac
6.506900
[0x004c6c7c]> ?t aar
1.181867
[0x004c6c7c]> ?t aan
0.087925
```

`aan` is a bit slower than on master, but `aar` is slightly faster.
`aac` instead is still much slower than master, I still need to analyze the cause.

**Comment by ret2libc** [comment]
With latest commits:
```
r2 ./radare2-regressions/bins/anal/keygen.exe
Warning: off > bin->size
Warning: off > bin->size
 -- Use 'e' and 't' in Visual mode to edit configuration and track flags.
[0x004c6c7c]> ?t aa
[x] Analyze all flags starting with sym. and entry0 (aa)
0.002168
[0x004c6c7c]> ?t aac
2.221273
[0x004c6c7c]> ?t aar
1.876399
[0x004c6c7c]> ?t aan
0.455914
```

we are back to good numbers :)

File: libr/anal/fcn.c [hunk 1]
```diff
@@ -1491,13 +1426,6 @@ static int fcn_recurse(RAnal *anal, RAnalFunction *fcn, ut64 addr, ut8 *buf, ut6
 			}
 			// switch statement
 			if (anal->opt.jmptbl) {
-				if (fcn->refs->tail) {
```
Comment (ret2libc) [inline libr/anal/fcn.c]: @Toizi @radare maybe I should not remove these lines, right? :)
What I don't understand though is... what is the "tail" of the refs? What is the "last" ref?

**Comment by radare** [comment]
woooooo :D

**Comment by radare** [comment]
i think this refs->tail code was about to check if there function had any ref or not

**Comment by radare** [comment]
aar was faster with dict, but definitvely faster than anything else in master. good work :)

File: libr/anal/bb.c [hunk 2]
```diff
@@ -144,7 +144,7 @@ R_API int r_anal_bb(RAnal *anal, RAnalBlock *bb, ut64 addr, ut8 *buf, ut64 len,
 					ut8 b[8];
 					ut64 ptr = idx+addr+src->delta;
 					anal->iob.read_at (anal->iob.io, ptr, b, memref);
-					r_anal_ref_add (anal, ptr, addr+idx-op->size, 'd');
+					r_anal_xrefs_set (anal, R_ANAL_REF_TYPE_DATA, addr+idx-op->size, ptr);
```
Comment (radare) [inline libr/anal/bb.c]: any reason to pass type before from/to?

File: libr/anal/fcn.c (hunk 1 above)
Comment (radare) [reply libr/anal/fcn.c]: i think fcn->refs->tail is an assumption to check  if there's any ref

File: libr/core/cmd_anal.c [hunk 3]
```diff
@@ -4748,14 +4748,14 @@ static void _anal_calls(RCore *core, ut64 addr, ut64 addr_end) {
 				if (!anal_is_bad_call (core, from, to, addr, buf, bufi)) {
 					fcn = r_anal_get_fcn_in (core->anal, op.jump, R_ANAL_FCN_TYPE_ROOT);
 					if (!fcn) {
-						r_core_anal_fcn (core, op.jump, addr, R_ANAL_REF_TYPE_NULL, depth);
```
Comment (radare) [inline libr/core/cmd_anal.c]: RCore.anal_fcn is having reftype after from/to, so make things consistent, revert the change in the xrefs api

File: libr/core/cmd_anal.c [hunk 4]
```diff
@@ -6143,7 +6133,7 @@ void _CbInRangeAav(RCore *core, ut64 from, ut64 to, int vsize, bool asterisk, in
 		r_cons_printf ("f+ aav.0x%08"PFMT64x "= 0x%08"PFMT64x, to, to);
 	} else {
 #if 1
-		r_anal_ref_add (core->anal, to, from, ' ');
+		r_anal_xrefs_set (core->anal, ' ', from, to);
```
Comment (radare) [inline libr/core/cmd_anal.c]: use the enum of R_ANAL_XREF_TYPE_ instead of ' '

File: libr/anal/xrefs.c [hunk 5]
```diff
@@ -24,166 +21,117 @@ xrefs
...
+}
+
 static const char *analref_toString(RAnalRefType type) {
 	switch (type) {
 	case R_ANAL_REF_TYPE_NULL:
 		/* do nothing */
 		break;
 	case R_ANAL_REF_TYPE_CODE:
-		return "code.jmp";
+		return "code jmp";
```
Comment (radare) [inline libr/anal/xrefs.c]: i think we can remove the code and data words in here, because jmp/call only makes sense for code and mem and string only for data., lets rename those enums too.

the problem i see is that TYPE_CODE is used for LEA when pointing to code and JMP, so cant be TYPE_JMP.. what do you think? we can keep those, but maybe remove the code/data to start

File: libr/anal/xrefs.c [hunk 6]
```diff
@@ -24,166 +21,117 @@ xrefs
...
-		eprintf ("%s 0x%08llx -> 0x%08llx (0x%llx)\n",
-				kv->u, kv->k, kv->v, addr);
-#endif
-		ref->at = k;
-		ref->addr = v;
-		ref->type = u;
-		r_list_append (list, ref);
-	}
+static bool appendRef(RList *list, const char *k, RAnalRef *ref) {
+	RAnalRef *cloned = r_anal_ref_new ();
```
Comment (radare) [inline libr/anal/xrefs.c]: check for null

File: libr/core/cmd_anal.c (hunk 4 above)
Comment (ret2libc) [reply libr/core/cmd_anal.c]: The problem is that there's no enum for ' '.
I don't know how that worked so far...

File: libr/anal/xrefs.c [hunk 7]
```diff
@@ -24,166 +21,117 @@ xrefs
...
+		ht_insert_u64 (m, from, ht);
 	}
-}
 
-// [from=[from:to,],]
-// 10->20
-static void setref(dict *m, ut64 from, ut64 to, int type) {
-	dict_set (m, from, to, r_anal_xrefs_type_tostring (type));
-}
+	RAnalRef *ref = r_anal_ref_new ();
```
Comment (radare) [inline libr/anal/xrefs.c]: chk for null

File: libr/anal/xrefs.c [hunk 8]
```diff
@@ -195,333 +143,131 @@ R_API int r_anal_xrefs_set (RAnal *anal, const RAnalRefType type, ut64 from, ut6
...
+	r_list_foreach (list, iter, ref) {
+		int t = ref->type ? ref->type: ' ';
+		switch (rad) {
+		case '*':
+			anal->cb_printf ("ax%c 0x%"PFMT64x" 0x%"PFMT64x"\n",
+				t, ref->at, ref->addr);
+			break;
+		case '\0':
+			{
+				char *name = anal->coreb.getNameDelta (anal->coreb.core, ref->at);
```
Comment (radare) [inline libr/anal/xrefs.c]: check for null

File: libr/anal/xrefs.c [hunk 9]
```diff
@@ -195,333 +143,131 @@ R_API int r_anal_xrefs_set (RAnal *anal, const RAnalRefType type, ut64 from, ut6
...
+				r_str_replace_char (name, ' ', 0);
+				anal->cb_printf ("%40s", name? name: "");
+				free (name);
+				anal->cb_printf (" 0x%"PFMT64x" -> %9s -> 0x%"PFMT64x, ref->at, analref_toString (t), ref->addr);
+				name = anal->coreb.getNameDelta (anal->coreb.core, ref->addr);
+				r_str_replace_char (name, ' ', 0);
```
Comment (radare) [inline libr/anal/xrefs.c]: null deref, you are checking for name being null after using it in r_str_replace_char()

File: libr/anal/xrefs.c [hunk 10]
```diff
@@ -537,20 +283,84 @@ R_API const char *r_anal_xrefs_type_tostring (char type) {
...
-	CountState cs = { anal, 0 };
-	sdb_foreach (DB, (SdbForeachCallback)countcb, &cs);
-	return cs.count;
+	return anal->dict_xrefs->count;
+}
+
+static int ref_cmp(const RAnalRef *a, const RAnalRef *b) {
+	if (a->at < b->at) {
+		return -1;
+	} else if (a->at > b->at) {
```
Comment (radare) [inline libr/anal/xrefs.c]: dont else after return

File: libr/anal/xrefs.c [hunk 11]
```diff
@@ -537,20 +283,84 @@ R_API const char *r_anal_xrefs_type_tostring (char type) {
...
+		return 1;
+	} else {
+		if (a->addr < b->addr) {
+			return -1;
+		} else if (a->addr > b->addr) {
```
Comment (radare) [inline libr/anal/xrefs.c]: same here

File: libr/anal/xrefs.c [hunk 12]
```diff
@@ -537,20 +283,84 @@ R_API const char *r_anal_xrefs_type_tostring (char type) {
...
+R_API RList *r_anal_fcn_get_refs(RAnal *anal, RAnalFunction *fcn) {
+	return fcn_get_refs (fcn, anal->dict_refs);
+}
+
+R_API RList *r_anal_fcn_get_xrefs(RAnal *anal, RAnalFunction *fcn) {
+	return fcn_get_refs (fcn, anal->dict_xrefs);
+}
+
+R_API RList *r_anal_fcn_get_refs_sorted(RAnal *anal, RAnalFunction *fcn) {
+	RList *l = r_anal_fcn_get_refs (anal, fcn);
```
Comment (radare) [inline libr/anal/xrefs.c]: this is a oneliner, dont waste 3 lines, imho return r_list_sort (r_anal_fcn_get_Refs (anal, fcn), ref_cmp) should be good. 

also the cast for the function signature shouldnt be needed if the function gets void*

File: libr/anal/xrefs.c (hunk 5 above)
Comment (ret2libc) [reply libr/anal/xrefs.c]: I'm fine with removing `code` and `data` prefixes. Though I would convert R_ANAL_REF_TYPE_CODE to "code" instead of "jmp", because as you said TYPE_CODE is not only used for jmps.

File: libr/anal/bb.c (hunk 2 above)
Comment (ret2libc) [reply libr/anal/bb.c]: I will adjust this.

File: libr/anal/xrefs.c (hunk 9 above)
Comment (ret2libc) [reply libr/anal/xrefs.c]: it was already there, I didn't touch nor looked at this code at all. I just assumed it was working, but i can change that in this PR too.

File: libr/anal/xrefs.c (hunk 12 above)
Comment (ret2libc) [reply libr/anal/xrefs.c]: 1) r_list_sort unfortunately doesn't return a RList*, so you can't really do it in one line
2) the cast for the function signature is needed because i want to directly use RAnalRef as parameters. If you do otherwise, the signature should be (void*,void*) and you need to cast *both* arguments inside the function, which seems useless when you can just cast the function signature, as I did.

File: libr/anal/xrefs.c (hunk 9 above)
Comment (ret2libc) [reply libr/anal/xrefs.c]: this is actually fine too, because r_str_replace_char checks for NULL.

File: libr/anal/xrefs.c (hunk 8 above)
Comment (ret2libc) [reply libr/anal/xrefs.c]: no need even here, because r_str_replace_char already checks NULL, cb_printf checks NULL and free works on NULL too

File: libr/anal/xrefs.c (hunk 5 above)
Comment (ret2libc) [reply libr/anal/xrefs.c]: So... weird idea but if instead of having this `analref_toString` function we just use `r_anal_xrefs_type_tostring`? it would be consistent everywhere at least!

File: libr/anal/xrefs.c (hunk 12 above)
Comment (radare) [reply libr/anal/xrefs.c]: oops right about the rlistSort() meh its fine yep i know about this caveat, but casting a function signature is more dangerous than casting the arguments because you can pass a wrong callback and didnt notice until running., if the funciton have different numnber of arguments or types it will complain at compile time. i can merge like this , it was just a comment :)

File: libr/anal/xrefs.c (hunk 9 above)
Comment (radare) [reply libr/anal/xrefs.c]: ok

File: libr/anal/xrefs.c (hunk 5 above)
Comment (radare) [reply libr/anal/xrefs.c]: yeah sounds good to me, if you want to do any other change here we can discuss it after merging. focus on pleasing travis and ill merge :))))

**Comment by ret2libc** [comment]
merged with https://github.com/radare/radare2/commit/b703dfbf17c0cff54181627fb40da92de3a52b79
